keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md

```diff
@@ -0,0 +1,437 @@
+   thus each has N pods.
+2. A failure or an upgrade happens in one of the domains. The N pods from this
+   domain get re-scheduled into the other 2 domains. Note that this N pods are
+   now the youngest.
+3. The domain recovers or finishes upgrading.
+4. ReplicaSet is downscaled to 2N, due to user action or HPA recommendation.
+   Given the downscaling algorithm, 2 domains end up with N nodes each, the 2N
+   Pods that were never restarted, and the remaining domain has 0 Pods.
+   There is nothing to be done here. A random approach would obtain the same
+   result.
+5. The ReplicaSet is upscaled to 3N again, due to user action or HPA
+   recommendation. Due to Pod spreading during scheduling, each domain has N
+   Pods. Balance is recovered. However, one failure domain holds the youngest
+   Pods.
+6. ReplicaSet is downscaled to 2N again. Due to the downscaling preference, all
+   the Pods from one domain are removed, leading to imbalance. The situation
+   doesn't improve with repeated upscale and downscale steps. Instead, a
+   randomized approach leaves about 2/3*N nodes in each
+   failure domain.
+
+### Notes/Constraints/Caveats
+
+The original heuristic, dowscaling the youngest Pods first, has its benefits.
+Newer Pods might not have finished starting up (or warming up) and are likely
+to have less active connections than older Pods. However, this distinction
+doesn't generally apply once Pods have been running steadily for some time.
+
+A purely randomized approach would break those assumptions, potentially
+leading to services disruption. Choosing a heuristic could be left to the user.
+On the other hand, certain workloads take a long time to warm up and, at the
+same time, require high availability.
+
+### Risks and Mitigations
+
+Certain users might be relaying in the existing downscaling heuristic. However,
+there are a number of reasons why we don't need to preserve such behavior as is:
+
+- The behavior is not documented in the API or user docs, only in code.
+- The heuristic is applied last, after other criteria have been applied. In
+  particular, the heuristic doesn't apply when multiple Pods are in the same
+  Node or if a particular Pod has had several container restarts. While users
+  can enforce one Pod per Node with certain scheduling features, there is no
+  workaround for other criteria.
+- We are not proposing to entirely remove the heuristic, just make it more
+  laxed.
+- We are introducing a [related feature](git.k8s.io/enhancements/keps/sig-apps/1828-delete-priority-annotations)
+  that provides higher guarantees for downscaling order that users can migrate
+  to.
+  
+  <!--
+  If not merged yet, the KEP can be found in https://github.com/kubernetes/enhancements/pull/1828
+  -->
+
+## Design Details
+
+We propose a randomized approach to the algorithm for Pod victim selection
+during ReplicaSet downscale:
+
+1. Do a random shuffle of ReplicaSet Pods.
```

Comment:
why is this necessary?

Reply:
I'm not sure if just quick sort, with the random pivot, provides the guarantees that any order is equally likely.
If it doesn't, then we would favor removing Pods in the order provided by the lister.

Reply:
the last sorting key can be a random number or the pods' uuid

Reply:
uid sounds good. The random number would have to be produced before calling sort (and not in the `Less` function) otherwise it leads to undefined behavior.

Reply:
Yes, the random numbers need to be assigned before hand, hence the uuid suggestion.

Reply:
Do we want to sort the pods before or after grouping them into their scale buckets? I thought the goal was to select randomly from the youngest bucket

Reply:
The goal is that Pods that belong to the same scale bucket are in a random order.

Maybe we don't need to get into the implementation detail of how we achive that in the KEP, but what it's important to note is what is the precedence of the buckets with regard to the other sorting criteria https://github.com/kubernetes/kubernetes/blob/cac9339/pkg/controller/controller_utils.go#L784-L809

Reply:
> The goal is that Pods that belong to the same scale bucket are in a random order.
> Maybe we don't need to get into the implementation detail

Agree, just making sure I understood the intent.